### PR TITLE
WIP: Add Invite Endpoints

### PR DESCRIPTION
**Pull request by AshGodfrey**
## Documentation

Does this require changes to the WorkOS Docs? E.g. the [API Reference](https://workos.com/docs/reference) or code snippets need updates.

```
[ ] Yes
```

If yes, link a related docs PR and add a docs maintainer as a reviewer. Their approval is required.

### CODE REVIEW

File: pkg/users/client.go [hunk 1]
```diff
@@ -238,6 +238,44 @@ type ListAuthFactorsResponse struct {
 	ListMetadata common.ListMetadata `json:"listMetadata"`
 }
 
+type InviteObject struct {
+	Object         string `json:"object"`
```
Comment (amygdalama) [inline pkg/users/client.go:242]:
```suggestion
type Invite struct {
```
small naming change + looks like we omit the object field for other object types.

File: pkg/users/client.go [hunk 2]
```diff
@@ -238,6 +238,44 @@ type ListAuthFactorsResponse struct {
...
+	ID             string `json:"id"`
+	Email          string `json:"email"`
+	State          string `json:"state"`
```
Comment (amygdalama) [inline pkg/users/client.go:245]: enumerate the different `state`s?
* pending
* accepted
* expired
* revoked

(not sure how to do that in go, hopefully we have some existing examples to follow...)

File: pkg/users/client.go [hunk 3]
```diff
@@ -238,6 +238,44 @@ type ListAuthFactorsResponse struct {
...
+	AcceptedAt     string `json:"accepted_at,omitempty"`
+	RevokedAt      string `json:"revoked_at,omitempty"`
+	Token          string `json:"token"`
+	OrganizationID string `json:"organization_id,omitempty"`
+	ExpiresAt      string `json:"expires_at"`
+	CreatedAt      string `json:"created_at"`
+	UpdatedAt      string `json:"updated_at"`
+}
+
+type CreateInvitationOpts struct {
```
Comment (amygdalama) [inline pkg/users/client.go:255]: we call these "invite" / "invites" in our API so i think we should be consistent w/ that in the SDKs

File: pkg/users/client.go [hunk 4]
```diff
@@ -238,6 +238,44 @@ type ListAuthFactorsResponse struct {
...
+type RevokeInvitationOpts struct {
+	InviteID string
+}
+
+type ListInvitationsOpts struct {
+	OrganizationID string `json:"organization_id,omitempty"`
+	Email          string `json:"email,omitempty"`
+}
+
+type GetInvitationByIDOpts struct {
```
Comment (amygdalama) [inline pkg/users/client.go:271]:
```suggestion
type GetInviteOpts struct {
```

looks like the corresponding types for other endpoints don't have a "by ID" suffix (like `GetUserOpts`). let's match that convention; it also matches the API.

File: pkg/users/client.go [hunk 5]
```diff
@@ -1066,3 +1104,174 @@ func (c *Client) ListAuthFactors(ctx context.Context, opts ListAuthFactorsOpts)
...
+
+	var body InviteObject
+	dec := json.NewDecoder(res.Body)
+	err = dec.Decode(&body)
+
+	return body, err
+}
+
+// ListInvitations gets a list of all of your existing invites matching the criteria specified.
+func (c *Client) ListInvitations(ctx context.Context, opts ListInvitationsOpts) ([]InviteObject, error) {
```
Comment (amygdalama) [inline pkg/users/client.go:1176]: what about the pagination part of the response?

**Review by amygdalama — COMMENTED** [review]
i only reviewed the types + responses (deferring correctness / go conventions to someone in the go group)

thank you for picking this up!! 🙏

File: pkg/users/client.go (hunk 2 above)
Comment (AshGodfrey) [reply pkg/users/client.go:245]: No worries - easy enough, I'll add these :)

File: pkg/users/client.go (hunk 4 above)
Comment (AshGodfrey) [reply pkg/users/client.go:271]: How does this work when we also have getInvitationByToken? It seems like token and ID should have separate options to pass in? Or are we using a "GetInviteOpts" with a token and ID for both?

Comment (AshGodfrey) [reply pkg/users/client.go:271]: Or is one `GetInviteOpts` and the other `GetInviteByTokenOpts`?

Comment (AshGodfrey) [reply pkg/users/client.go:271]: We usually use the same name as the function. So if it's "`GetInvite`" for ID and "`GetInviteByToken`" for token this makes sense. However if we want `GetInviteByID`, structurally we probably want to go with `GetInviteByIDOpts`

Comment (amygdalama) [reply pkg/users/client.go:271]: yeah i was thinking two separate functions `GetInvite` and `GetInviteByToken` (which matches the API implementation). i think as a default we should match the API implementation to keep it straightforward, and then we can make exceptions in cases where the API is inconsistent or not idiomatic. do you think this is one of those cases?

Comment (AshGodfrey) [reply pkg/users/client.go:271]: Cool as long as it's consistent, I'm 100% for it - will push these changes today.

File: pkg/users/client.go [hunk 6]
```diff
@@ -26,6 +26,17 @@ const (
 	Desc Order = "desc"
 )
 
+// TokenState represents the state of a token.
+type TokenState string
```
Comment (amygdalama) [inline pkg/users/client.go:30]: call this `InviteState` instead pls!

File: pkg/users/client.go [hunk 7]
```diff
@@ -238,6 +249,65 @@ type ListAuthFactorsResponse struct {
...
+	// Pagination cursor to receive records before a provided User ID.
+	Before string `url:"before,omitempty"`
+
+	// Pagination cursor to receive records after a provided User ID.
+	After string `url:"after,omitempty"`
+}
+
+type GetInviteOpts struct {
+	InviteID string
+}
```
Comment (amygdalama) [inline pkg/users/client.go:305]:
```suggestion
type GetInviteOpts struct {
        // Invite unique identifier
	Invite string `json:"id"`
}
```
looks like in other `Get*Opts` cases we use the object name here instead, so let's follow that convention:
https://github.com/workos/workos-go/blob/c831fd777f3fa50f092ab4cb880bb768f33a3c14/pkg/users/client.go#L86-L90

(sorry i didn't catch this before)

Comment (amygdalama) [reply pkg/users/client.go:305]: idk if `json:"id"` is correct btw i just copy/pasta-ed from the user case

Comment (amygdalama) [reply pkg/users/client.go:305]: and i suggest using that same convention for `RevokeInviteOpts`

Comment (AshGodfrey) [reply pkg/users/client.go:305]: yeah this is a contested naming convention at times, sometimes we call it `User` other times we call it `UserID` or just `ID` (this particular package is going with the `User` style convention but we've debated it in slack). I'll go ahead and call it `User` style for package consistency

Comment (AshGodfrey) [reply pkg/users/client.go:305]: Also it doesn't need the JSON string here because it's being passed in in the URL.

Comment (AshGodfrey) [reply pkg/users/client.go]: JSON tags are irrelevant in the opts - just the response where there's JSON encoding/decoding.

**Review by amygdalama — APPROVED** [review]
names + types look good to me!

will defer to someone from the go group on implementation specifics

**Comment by amygdalama** [comment]
hold up on this; we're making some last-minute changes to the API

**Pull request closed without merging**